Clean up handling of 'taskType' field

tsmaeder opened the pull request:
Signed-off-by: Thomas Mäder <tmader@redhat.com>

<!--
Thank you for your Pull Request. Please provide a description and review
the requirements below.

Contributors guide: https://github.com/theia-ide/theia/blob/master/CONTRIBUTING.md
-->

#### What it does
This PR cleans up a couple of the issues introduced by Implement CustomExecution extension API https://github.com/eclipse-theia/theia/pull/9189 while retaining the custom execution API.

The approach is to strictly use the "taskType" field in a task configuration to mean the "task execution type". The magic in types-impl.ts (`Task#updateTypeBasedOnExecution`) has been removed and the translation of `Task.execution` into `TaskDTO.taskType` and vice-versa has been moved to the type converter.

I have removed a couple of instances where `taskType` was used erroneously when the declared type was meant and not the execution type.

I have removed some filtering of dependent tasks based on "source".  It did not work (configured tasks have the folder they are in as "source") and I don't think the `source` field should be considered part of the task identity.

I have also removed the "garbage collection" of callback functions for custom executions. There is no explicit lifecycle for those callback functions that I could find in the VS Code API. I believe there are supposed to be only a fixed, small number of callback functions. So I'm retaining these functions based on identity. We could clean up the callbacks whenever we call `provideTasks` and after call to the `executeTask` API, but I don't think it is necessary. Convince me otherwise :-)

#### How to test
<!-- Explain how a reviewer can reproduce a bug, test new functionality or verify performance improvements. -->
Make sure non-contributed tasks still work, aka `process` and `shell` tasks.

Make sure we can run and configure both custom execution tasks (see gradle tasks: https://github.com/eclipse-theia/theia/issues/8767) and the shell tasks contributed by @vince-fugnitto's test plugin (see https://github.com/eclipse-theia/theia/issues/9366)

Make sure we can run dependent tasks based on label as well as based on structured task identifier.

#### Review checklist

- [x] as an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)

#### Reminder for reviewers

- as a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

RomanNikitenko reviewed (commented):
@tsmaeder 
thank you for the provided fix, I tested the changes and found that they don't fix https://github.com/eclipse-theia/theia/pull/9207#pullrequestreview-637691473.
It looks like we still have some problem with `provided` tasks:

![test_task_bug](https://user-images.githubusercontent.com/5676062/115390886-7822dc80-a1e7-11eb-88da-8267223c33cf.gif)

The use case works well with these changes https://github.com/eclipse-theia/theia/pull/9373:

![test_task_no_bug](https://user-images.githubusercontent.com/5676062/115391187-db147380-a1e7-11eb-8eef-5b1de14d60e2.gif)

RomanNikitenko commented:
Also, there is still a problem from https://github.com/eclipse-theia/theia/issues/9365.
It's number 3 in https://github.com/eclipse-theia/theia/issues/9365:

> It looks like after configuring a task the configuration from tasks.json file is not taken into account at all - so a user select a configured configuration for running, but the corresponding detected configuration is used instead.

![configured_task_bug](https://user-images.githubusercontent.com/5676062/115393861-d30a0300-a1ea-11eb-95f6-a46193df8755.gif)

So, I have configured a task, changed the configuration, but task system still run the corresponding `detected` task, not configured task...

tsmaeder commented:
@RomanNikitenko there is https://github.com/eclipse-theia/theia/pull/9207/files for the problem matchers.

RomanNikitenko commented:
> @RomanNikitenko there is https://github.com/eclipse-theia/theia/pull/9207/files for the problem matchers.

What do you mean?
is it related to my first comment https://github.com/eclipse-theia/theia/pull/9377#pullrequestreview-639847731?
If so - I have checked that behavior for your changes and the same for these changes https://github.com/eclipse-theia/theia/pull/9373

tsmaeder commented:
> is it related to my first comment #9377 (review)?

Yes, there is an open PR for that issue.

RomanNikitenko commented:
> > is it related to my first comment #9377 (review)?
> 
> Yes, there is an open PR for that issue.

I don't understand what do you mean.
I tested the use case described here https://github.com/eclipse-theia/theia/pull/9207#pullrequestreview-637691473 using:
- your changes - it doesn't work
- theses changes https://github.com/eclipse-theia/theia/pull/9373 - it works well

tsmaeder commented:
@RomanNikitenko so you're doing this:

1. Configure the "tsc: build" task in a workspace containing theia source
2. Run the task.
3. Observe you get output in a terminal named 'npm'
4. Add a "presentation" object to the configuration
5. Change "panel" to "new" 
6. Change "echo" to "false"
7. Run the task
8. Observe: the output of the command is shown in the same terminal as before

You expectation would be that there is no output or that there is a new terminal being opened and that the command line being executed is not shown in the terminal?

RomanNikitenko commented:
> @RomanNikitenko so you're doing this:
> 
> 1. Configure the "tsc: build" task in a workspace containing theia source
> 2. Run the task.
> 3. Observe you get output in a terminal named 'npm'
> 4. Add a "presentation" object to the configuration
> 5. Change "panel" to "new"
> 6. Change "echo" to "false"
> 7. Run the task
> 8. Observe: the output of the command is shown in the same terminal as before
> 
> You expectation would be that there is no output or that there is a new terminal being opened and that the command line being executed is not shown in the terminal?

My expectation is: the configured task (not detected) is running when I run `configured` task

colin-grant-work commented on packages/plugin-ext/src/plugin/tasks/tasks.ts:
```diff
@@ -49,9 +52,9 @@ export class TasksExtImpl implements TasksExt {
-        this.providedCustomExecutions = new Map<number, CustomExecution>();
-        this.notProvidedCustomExecutions = new Set<number>();
-        this.activeCustomExecutions = new Map<number, CustomExecution>();
+        this.customExecutionIds = new Map();
```
[Minor]: some map fields are initialized as properties above (`adaptersMap`, `executions`); others are initialized here. I'd vote for moving these three, which don't need information from the constructor arguments, up with the rest and initializing with the property syntax.

colin-grant-work commented on packages/plugin-ext/src/main/browser/tasks-main.ts:
```diff
@@ -234,6 +234,7 @@ export class TasksMainImpl implements TasksMain, Disposable {
         return {
             ...common,
             ...partialDto,
+            taskType: taskType,
```
[Minor]: Could be just `taskType`.

RomanNikitenko replied:
or https://github.com/eclipse-theia/theia/pull/9377#discussion_r617070847

RomanNikitenko replied:
so, it looks like we don't need the change here

colin-grant-work reviewed (commented):
I'll need a bit of time to get a better understanding of all of the contingencies in play, but here are a couple of very minor comments from my first read-through of the changes.

tsmaeder commented:
@RomanNikitenko 

> My expectation is: the configured task (not detected) is running when I run configured task

We _are_ running the configured task. However, we pass the configured task to `taskProvider.resolve()` of the typescript built-in extension, which returns a new task which has an empty `presentation` object. 

I suspect this "worked" before because we were using the "taskType" based on the task execution (which is `process` or `shell`) to look up the task resolver. We would then not get the plugin task resolver, but the fallback "ProcessTaskResolver", which doesn't empty the presentation field.

tsmaeder commented:
Turns out that VSCode overrides the configured properties after calling the task resolver: https://github.com/microsoft/vscode/blob/10c56b598de01119b80a5c53871003f868fd41fb/src/vs/workbench/parts/tasks/node/taskConfiguration.ts#L1473

alvsan09 commented:
I found the following issue when testing with;
https://github.com/microsoft/vscode-extension-samples/tree/main/task-provider-sample

The output from custom task e.g. task 'custombuildscript: 32 -incremental' does not show the extensions output in the terminal, 
you can compare the expected and actual results in the following images when running the task twice in a row.

![task32_incremental](https://user-images.githubusercontent.com/76971376/115429934-e65ca480-a1d1-11eb-81be-09332194bd03.png)

![task-32-incremental-no-output](https://user-images.githubusercontent.com/76971376/115429955-ebb9ef00-a1d1-11eb-8152-970317963f53.png)

alvsan09 commented:
I have tried @RomanNikitenko case but using "gradle: clean", I updated the presentation options under the configured task (tasks.json) and confirmed that the presentation options don't take effect.

alvsan09 commented on packages/plugin-ext/src/main/browser/tasks-main.ts:
```diff
@@ -219,7 +219,7 @@ export class TasksMainImpl implements TasksMain, Disposable {
 
     protected fromTaskConfiguration(task: TaskConfiguration): TaskDto {
-        const { group, presentation, _scope, _source, ...common } = task;
+        const { group, presentation, _scope, _source, taskType, ...common } = task;
```
taskType does not seem to be converted, so we can leave it as part of ...comon, right ?

alvsan09 reviewed (commented):
Nice improvement !
it would be great to rename all taskType instances that mean execution type, so taskType names that refer to actual TaskDefintion.type would not be confused with execution type.

tsmaeder commented:
`TaskProvider.resolveTask()` is not invoked with the changes in https://github.com/eclipse-theia/theia/pull/9373/files, so it's broken in both cases.

tsmaeder commented:
I checked that in master, the presentation options of the configured `tsc: build` task are not used either. So at least it's no worse.

tsmaeder commented:
> I found the following issue when testing with;
https://github.com/microsoft/vscode-extension-samples/tree/main/task-provider-sample

@alvsan09 Sorry, but I don't understand what issue you're describing here: could you give steps to reproduce?

RomanNikitenko commented:
> I checked that in master, the presentation options of the configured `tsc: build` task are not used either.

I believe it was working well before merging https://github.com/eclipse-theia/theia/pull/9189.

We have a lot of regressions in master after merging https://github.com/eclipse-theia/theia/pull/9189:
- https://github.com/eclipse-theia/theia/issues/9341
- https://github.com/eclipse-theia/theia/issues/9365
- https://github.com/eclipse-theia/theia/issues/9366
- https://github.com/eclipse-theia/theia/pull/9207#pullrequestreview-637691473

tsmaeder commented:
I believe all of those are fixed, except https://github.com/eclipse-theia/theia/pull/9207#pullrequestreview-637691473 (for which there is a separate PR already open).

As for the presentation options not being used, IMO that only ever worked because `TaskProvider#resolve()` was never invoked. So it was not working, just broken in a different way.

see https://github.com/eclipse-theia/theia/pull/9377#issuecomment-823820207

alvsan09 commented:
> > I found the following issue when testing with;
> > https://github.com/microsoft/vscode-extension-samples/tree/main/task-provider-sample
> 
> @alvsan09 Sorry, but I don't understand what issue you're describing here: could you give steps to reproduce?

* Install the extension mentioned above
* In theia (any project) run the plugin provided task 'custombuildscript: 32 incremental' i.e. Terminal -> Run task ...
* run it twice and see that the output in terminal is not complete i.e. it does not show the expected plugin provided messages highlighted in the red square.

![image](https://user-images.githubusercontent.com/76971376/115554553-938bf700-a27c-11eb-91bd-039c22e0fa62.png)

alvsan09 commented:
> As for the presentation options not being used, IMO that only ever worked because `TaskProvider#resolve()` was never invoked. So it was not working, just broken in a different way.
> 
@tsmaeder, could you please clarify what you mean with it was broken in a different way ?
From a user's perspective it's a regression, right ?

tsmaeder commented on packages/plugin-ext/src/common/plugin-api-rpc.ts:
```diff
@@ -1354,6 +1354,8 @@ export interface CommandProperties {
 
 export interface TaskDto {
     type: string;
+    taskType?: 'shell' | 'process' | 'customExecution'; // the task execution type
```
I'm trying to minimize the changes in this one. Renaming can be done separately.

alvsan09 replied:
you are right, it's better to leave it for a later more specific change.

tsmaeder commented:
> @tsmaeder, could you please clarify what you mean with it was broken in a different way ?
> From a user's perspective it's a regression, right ?

Well, if you cared about configuring the presentation options, it was a regression, but if you cared about `TaskProvider.resolve()` bein called, it was a bugfix :-) But anyway...with my latest commits, I believe both codepaths are covered now.

alvsan09 commented:
> > @tsmaeder, could you please clarify what you mean with it was broken in a different way ?
> > From a user's perspective it's a regression, right ?
> 
> Well, if you cared about configuring the presentation options, it was a regression, but if you cared about `TaskProvider.resolve()` bein called, it was a bugfix :-) But anyway...with my latest commits, I believe both codepaths are covered now.

I can confirm that the presentation options work for me taken the latest branch of this PR

tsmaeder commented:
I've added two commits that do the following:

1. Reapply task customisations after calling `TaskProvider.resolve()`. This should take care of the case where customising a `tsc: build` task was not working.
2. Introduced the concept of the `TaskExcecutionResolver`. This came from the observation that we want to apply the `ProcessTaskResolver` to any task that has a `shell` or `process` task execution, not just the ones that have the `shell|process` task type. `ProcessTaskResolver` is responsible for doing stuff like substituting variables in for the `cwd` option, for example.

To test the latest change, try configuring the `cwd` option for a `tsc: build` task.

I'm re-requesting reviews, since the latest two changes are non-trivial and should be looked at.

colin-grant-work commented on packages/task/src/browser/task-service.ts:
```diff
@@ -726,27 +726,17 @@ export class TaskService implements TaskConfigurationClient {
      */
     getTaskByTaskIdentifierAndTaskDefinition(taskDefinition: TaskDefinition | undefined, taskIdentifier: TaskIdentifier, tasks: TaskConfiguration[]): TaskConfiguration {
```
The filter procedure here is a little opaque, and since we end up taking the first task that passes, it's really a `.find`. It might be a little clearer (and a little more efficient, I think), to do this:

```typescript
const requiredProperties = Object.keys(taskIdentifier);
const taskWithAllProperties = tasks.find(task => requiredProperties.every(property => task.hasOwnProperty(property) && task[property] === taskIdentifier[property]));
```
For each task, that will still bail out at the first property it fails on, and it makes a little clearer what we're checking for, I think. Then we can just

```typescript
return taskWithAllProperties ?? {label: '', _scope: '', type: ''}; // Fall back to empty TaskConfiguration
```

colin-grant-work commented on packages/task/src/browser/task-service.ts:
```diff
@@ -804,24 +794,36 @@ export class TaskService implements TaskConfigurationClient {
+            const resolver = await this.taskResolverRegistry.getTaskResolver(task.type);
+            const resolvedTask = resolver ? await resolver.resolveTask(task) : task;
+            const excutionResolver = this.taskResolverRegistry.getExecutionResolver(resolvedTask.taskType || resolvedTask.type);
+            overridePropertiesFunction(resolvedTask);
```
Do we need to call this function twice? If we call it here again after its resolved, under what circumstances is it necessary to have called it before resolution? And why call it here, after the first resolution and after checking for an execution resolver, but before running it through the execution resolver?

tsmaeder replied:
Yes, the double invocation is necessary: in particular, contributed tasks of type `npm` lose their presentation options when `TaskProvider.resolveTask()` is called. Also, VS Code seems to do the same (from reading their implementation).

RomanNikitenko replied:
Sorry, I'm not sure I fully understand this logic with double invocation.
How it worked before your changes without double invocation?

tsmaeder replied:
It did not work. That's why we were losing the presentation options ('shared' vs. 'new' terminal) with the `tsc: build` tasks. It kind of "worked" before the custom execution PR because `TaskProvider.resolve(task)` was simply never executed for any contributed task.

RomanNikitenko replied:
anyway it looks very strange that we have to override properties:
- before `await resolver.resolveTask(task)`
- and 
- after `await resolver.resolveTask(task)`

RomanNikitenko replied:
what's the goal to do it **before**, for example?

tsmaeder replied:
"resolve()" is never supposed to be called for purely contributed tasks. So there are always customisations whenever we call resolve for a task. These customisations might influence what a plugin does during the 'resolve' call. VS code seems to do the same thing 🤷

RomanNikitenko replied:
> resolve()" is **never** supposed to be called for purely contributed tasks. So there are **always** customisations

I don't think it's true - I can have a custom task type, custom resolver for it as well. I expect that my tasks are resolved using my resolver even if there are no customization for them.

Good example - some time ago we provided `che` tasks as `provided` or `detected` - we expected that they were resolved using `che` resolver even if there were no customization.
Now `che` tasks come as `configured` and we use `che` resolver as well. 
So  - `che` tasks as `detected` or as `configured` - we expect resolving them using our resolver - so I don't see some relations here...

RomanNikitenko replied:
In general - it's OK for me to override properties twice if we really know what's the reason

tsmaeder replied:
> I don't think it's true

It might not be true, but it should be according to the VS Code doc: https://code.visualstudio.com/api/references/vscode-api#TaskProvider

RomanNikitenko replied:
what exactly you mean in the doc?

tsmaeder replied:
Look at the documentation of the "resolveTask()" method of TaskProvider: 

> This method will not be called for tasks returned from the above provideTasks method since those tasks are always fully resolved.

RomanNikitenko replied:
thanks!
Then
> It kind of "worked" before the custom execution PR because TaskProvider.resolve(task) was simply never executed for any contributed task type.

it was the correct behavior and now it's broken - right?

tsmaeder replied:
No, the the resolve task should be called for customized task of any type that is not "shell" or "process". On the other hand, the resolution of variables should be called for any task that has an _execution_ of type `ShellExecution` or `ProcessExecution`. That behavior has never been implemented. What we have now is that we might call resolve on tasks where it is not necessary. I think that behaviour is superior to what we had before. This PR will not fix every problem in the Task system (that's why it's a roadmap item).

colin-grant-work commented:
Since @RomanNikitenko and @alvsan09 seem to have thee testing side down, I've mostly read over the code and made a few comments. In general, the issues that this fixes seem to me to call out for testing: the basic problem is that we do a set of transformations on data objects, and we want to confirm that the right thing comes out in the end. Would it be practical to mock up a couple of resolvers and run the `TaskService` through the steps that have been causing problems?

tsmaeder commented on packages/task/src/browser/task-contribution.ts:
```diff
@@ -96,38 +96,93 @@ export class TaskResolverRegistry {
+    /**
+     * Registers the given {@link TaskResolver} to resolve the `TaskConfiguration` for the
+     * specified type of execution ('shell', 'process' or 'customExecution').
+     * If there is already a `TaskResolver` registered for the specified type the registration will
+     * be overwritten with the new value.
```
Good point

RomanNikitenko commented on packages/task/src/browser/task-contribution.ts:
```diff
@@ -96,38 +96,93 @@ export class TaskResolverRegistry {
-    register(type: string, resolver: TaskResolver): Disposable {
-        this.resolvers.set(type, resolver);
+
+    registerTaskResolver(type: string, resolver: TaskResolver): Disposable {
```
Could you clarify why we need this method?
We already have a method to register task resolver with name `register` and `type: string, resolver: TaskResolver` parameters.
From my point of view  `register(type: string, resolver: TaskResolver)` within `TaskResolverRegistry` class is clear enough. 
You are introducing the method with the same parameters, but different name - what's the goal?

RomanNikitenko replied:
I guess it's because you are introducing another type of the resolver - `ExecutionResolver`, correct?

tsmaeder replied:
It's a rename: the original method is still there, but deprecated and to be removed. If there are two different `register...` methods, calling one `register` does not make sense.

RomanNikitenko commented on packages/task/src/browser/task-contribution.ts:
```diff
@@ -96,38 +96,93 @@ export class TaskResolverRegistry {
+     *
+     * @returns a promise of the registered `TaskResolver` or `undefined` if no resolver is registered for the given type.
+     */
+    async getTaskResolver(type: string): Promise<TaskResolver | undefined> {
```
Here I have the same opinion as I mentioned https://github.com/eclipse-theia/theia/pull/9377/files#r620049849:
`getResolver(type: string)` method within ` TaskResolverRegistry` class is readable enough, I believe we shouldn't introduce the new method.

tsmaeder replied:
And the same response :-)

RomanNikitenko commented on packages/task/src/browser/task-service.ts:
```diff
@@ -726,27 +726,17 @@ export class TaskService implements TaskConfigurationClient {
     getTaskByTaskIdentifierAndTaskDefinition(taskDefinition: TaskDefinition | undefined, taskIdentifier: TaskIdentifier, tasks: TaskConfiguration[]): TaskConfiguration {
         const identifierProperties: string[] = [];
-        let relevantTasks = tasks.filter(t =>
-            taskDefinition && t.hasOwnProperty('taskType') &&
```
you are removing using `taskDefinition` from the `getTaskByTaskIdentifierAndTaskDefinition()` method.
I believe the parameter `taskDefinition` should be removed and the documentation for the method should be updated

RomanNikitenko commented on packages/task/src/browser/task-service.ts:
```diff
@@ -804,24 +794,36 @@ export class TaskService implements TaskConfigurationClient {
+        try {
+            const resolver = await this.taskResolverRegistry.getTaskResolver(task.type);
+            const resolvedTask = resolver ? await resolver.resolveTask(task) : task;
+            const excutionResolver = this.taskResolverRegistry.getExecutionResolver(resolvedTask.taskType || resolvedTask.type);
```
ex**e**cutionResolver

RomanNikitenko commented:
I tested the changes and can confirm that they fix the following issues for me:
- https://github.com/eclipse-theia/theia/issues/9341
- https://github.com/eclipse-theia/theia/issues/9365
- https://github.com/eclipse-theia/theia/issues/9366

I sill not able to run a task according to https://github.com/eclipse-theia/theia/pull/9207#pullrequestreview-637691473
The error is: ```task ERROR Error resolving task 'test': TypeError: Cannot read property 'group' of undefined ```

![resolving_tasks](https://user-images.githubusercontent.com/5676062/116195835-65eef400-a73b-11eb-9190-2106c3f60cda.gif)

RomanNikitenko reviewed (commented):
I tested the changes and can confirm that they fix the following issues for me:
- https://github.com/eclipse-theia/theia/issues/9341
- https://github.com/eclipse-theia/theia/issues/9365
- https://github.com/eclipse-theia/theia/issues/9366

But I still not able to run a task according to https://github.com/eclipse-theia/theia/pull/9207#pullrequestreview-637691473
The error is: ```task ERROR Error resolving task 'test': TypeError: Cannot read property 'group' of undefined ```

![resolving_tasks](https://user-images.githubusercontent.com/5676062/116195835-65eef400-a73b-11eb-9190-2106c3f60cda.gif)

alvsan09 commented:
> > > I found the following issue when testing with;
> > > https://github.com/microsoft/vscode-extension-samples/tree/main/task-provider-sample
> > 
> > 
> > @alvsan09 Sorry, but I don't understand what issue you're describing here: could you give steps to reproduce?
> 
>     * Install the extension mentioned above
> 
>     * In theia (any project) run the plugin provided task 'custombuildscript: 32 incremental' i.e. Terminal -> Run task ...
> 
>     * run it twice and see that the output in terminal is not complete i.e. it does not show the expected plugin provided messages highlighted in the red square.
> 
> 
> ![image](https://user-images.githubusercontent.com/76971376/115554553-938bf700-a27c-11eb-91bd-039c22e0fa62.png)

Any news on this issue ?

vince-fugnitto commented:
@tsmaeder with the release tentatively scheduled for tomorrow (29th), how far are we from fixing regressions caused by https://github.com/eclipse-theia/theia/pull/9189? Should we think about preparing a revert, or perhaps delaying the release?

cc @marcdumais-work

tsmaeder commented:
IMO, the PR as it stands now is closer to a correct implementation than reverting the change.

tsmaeder commented:
> Any news on this issue ?

The problem here is that we're not registering custom execution callbacks for custom executions added during the `resolveTask` phase. I have pushed a fix.

alvsan09 commented:
> > Any news on this issue ?
> 
> The problem here is that we're not registering custom execution callbacks for custom executions added during the `resolveTask` phase. I have pushed a fix.

I can confirm that the fix is for this issue works !!, Thanks Thomas !

alvsan09 reviewed (approved):
I see only a pending issue for  #9207 which is still under review
I know my vote does not count for the merge, but I would like to reflect my opinion
as this fixes many outstanding issues present in master.

RomanNikitenko commented:
**As workaround** for the use case https://github.com/eclipse-theia/theia/pull/9377#pullrequestreview-645482849
we could consider these changes https://github.com/eclipse-theia/theia/commit/2aa654df5e117a271983acebb0a4e5a05374eb7a

You can use [my branch](https://github.com/eclipse-theia/theia/compare/fixTaskResolving?expand=1)  for testing - it contains all commits from the current PR + my workaround.

tsmaeder commented:
Just by the way...the doc for `resolveTask` also says that: 

> Note that when filling in the properties of task, you must be sure to use the exact same TaskDefinition and not create a new one. Other properties may be changed.

So the implementation in the task provider sample from the vscode-examples project is actually bogus: it returns new task objects.

RomanNikitenko commented:
> Just by the way...the doc for `resolveTask` also says that:
> 
> > Note that when filling in the properties of task, you must be sure to use the exact same TaskDefinition and not create a new one. Other properties may be changed.
> 
> So the implementation in the task provider sample from the vscode-examples project is actually bogus: it returns new task objects. The fix still applies, though: customized tasks may have their execution replaced.

Do you mean that there is no problem for [the use case](https://github.com/eclipse-theia/theia/pull/9377#pullrequestreview-645482849) if it it returns:
> the exact same TaskDefinition and not create a new one

?

Sorry, I didn't have a lot of time for a deep investigation, but for me it looks like it's a separate problem.
I found that `resolveTask` method can return `undefined` and tasks system doesn't take it in account.
I provided a fix for that.

It's true that there is a note in the VS Code documentation 
> > Note that when filling in the properties of task, you must be sure to use the exact same TaskDefinition and not create a new one. Other properties may be changed.

But for me it looks like it's not related to `undefined` and the fix that I provided.
Please be more precise, what influence that has within Theia tasks system.
I'll really appreciate if you provide some steps to reproduce or a use case.

tsmaeder commented:
@RomanNikitenko it's a completely unrelated problem. I just wanted to point out that we should not take the vscode samples as our bible: we should stick to what the documentation says, not what the samples do.

RomanNikitenko commented:
> @RomanNikitenko it's a completely unrelated problem. I just wanted to point out that we should not take the vscode samples as our bible: we should stick to what the documentation says, not what the samples do.

I see, thank you!

RomanNikitenko reviewed (approved):
I tested the changes and can confirm that they fix the following issues for me:
- https://github.com/eclipse-theia/theia/issues/9341
- https://github.com/eclipse-theia/theia/issues/9365
- https://github.com/eclipse-theia/theia/issues/9366

I'm able to run a task according to https://github.com/eclipse-theia/theia/pull/9207#pullrequestreview-637691473 using current PR changes + my changes: https://github.com/eclipse-theia/theia/commit/5d1efe335a0635e2d01eda51b61464b8027bb2eb, please see https://github.com/eclipse-theia/theia/pull/9377#issuecomment-828964655

tsmaeder commented:
@colin-grant-work @RomanNikitenko @alvsan09 if no-one has any last minute objections, I would go ahead and clean up and merge this one then. Thanks for the reviews & testing.

colin-grant-work commented:
> @colin-grant-work @RomanNikitenko @alvsan09 if no-one has any last minute objections, I would go ahead and clean up and merge this one then. Thanks for the reviews & testing.

I think that sounds good to me, as long as @RomanNikitenko code suggestions are taken into account.

RomanNikitenko commented:
> as long as @RomanNikitenko code suggestions are taken into account.

Do you mean these changes https://github.com/eclipse-theia/theia/commit/5d1efe335a0635e2d01eda51b61464b8027bb2eb ?
if so - they were not reviewed/tested.

I can not open a separate PR for them as the changes depend on the current PR.
So, maybe the true way is:
- squash and merge the current PR
- I align my branch with master
- I open the PR for https://github.com/eclipse-theia/theia/commit/5d1efe335a0635e2d01eda51b61464b8027bb2eb
- review ^
- merge if it looks good 

If you mean something else or have a better way - please let me know

vince-fugnitto commented:
> * squash and merge the current PR
> * I align my branch with master
> * I open the PR for [5d1efe3](https://github.com/eclipse-theia/theia/commit/5d1efe335a0635e2d01eda51b61464b8027bb2eb)
> * review ^
> * merge if it looks good

@RomanNikitenko @tsmaeder I propose we go forward with that approach, let's squash the current pull-request and merge (for today's release) and continue with 5d1efe3 in a separate pull-request.
